packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -362,49 +549,51 @@ class TernaryGadget : public GadgetT
     }
 };
 
-// (A <(=) B)
-class LeqGadget : public GadgetT
+// b ? A : B
+class ArrayTernaryGadget : public GadgetT
 {
 public:
-    VariableT _lt;
-    VariableT _leq;
-    libsnark::comparison_gadget<ethsnarks::FieldT> comparison;
 
-    LeqGadget(
+    std::vector<TernaryGadget> results;
+    VariableArrayT res;
+
+    ArrayTernaryGadget(
         ProtoboardT& pb,
-        const VariableT& A,
-        const VariableT& B,
-        const size_t n,
+        const VariableT& b,
+        const VariableArrayT& x,
+        const VariableArrayT& y,
         const std::string& prefix
     ) :
-        GadgetT(pb, prefix),
-
-        _lt(make_variable(pb, 1, FMT(prefix, ".lt"))),
-        _leq(make_variable(pb, 1, FMT(prefix, ".leq"))),
-        comparison(pb, n, A, B, _lt, _leq, FMT(prefix, ".A <(=) B"))
-    {
-        // The comparison gadget is only guaranteed to work correctly on values in the field capacity - 1
-        assert(n <= NUM_BITS_FIELD_CAPACITY - 1);
-    }
-
-    const VariableT& lt() const
+        GadgetT(pb, prefix)
     {
-        return _lt;
+        assert(x.size() == y.size());
+        results.reserve(x.size());
+        for (unsigned int i = 0; i < x.size(); i++)
+        {
+            results.emplace_back(TernaryGadget(pb, b, x[i], y[i], FMT(prefix, ".results")));
+            res.emplace_back(results.back().result());
+        }
     }
 
-    const VariableT& leq() const
+    void generate_r1cs_witness()
     {
-        return _leq;
+        for (unsigned int i = 0; i < results.size(); i++)
+        {
+            results[i].generate_r1cs_witness();
+        }
     }
 
-    void generate_r1cs_witness()
+    void generate_r1cs_constraints()
     {
-        comparison.generate_r1cs_witness();
+        for (unsigned int i = 0; i < results.size(); i++)
+        {
+            results[i].generate_r1cs_constraints(false);
+        }
     }
 
-    void generate_r1cs_constraints()
+    const VariableArrayT& result() const
     {
-        comparison.generate_r1cs_constraints();
+        return res;
```

Comment:
Our approach is to do an AND operator for each element, the approach above is to do a Sum operation in one constraint.

Reply:
Yes, that's why I put this in the comments:
> // This can be done more efficiently but we never have any long inputs so no need

We never do AND/OR on more than a couple of inputs. so need to make things more complicated than necessary. In general I'm not trying to do things as efficient as possible for these things because they are rarely used and are cheap either way.